crates/wash-lib/src/deps.rs

```diff
@@ -0,0 +1,408 @@
+//! Utilities for working with and managing wit dependencies
+
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use anyhow::{bail, Context, Result};
+use async_compression::tokio::bufread::GzipDecoder;
+use futures::TryStreamExt as _;
+use semver::VersionReq;
+use url::Url;
+use walkdir::WalkDir;
+use wasm_pkg_client::{
+    caching::{CachingClient, FileCache},
+    PackageRef,
+};
+use wasm_pkg_core::{config::Override, lock::LockFile, wit::OutputType};
+use wasmcloud_core::parse_wit_package_name;
+
+use crate::{
+    cli::CommonPackageArgs,
+    common::{clone_git_repo, RepoRef},
+    parser::{RegistryPullConfig, RegistryPullSource, RegistryPullSourceOverride},
+};
+
+/// Wrapper around a `wasm_pkg_client::Client` including configuration for fetching WIT dependencies.
+/// Primarily enables reuse of functionality to override dependencies and properly setup the client.
+pub struct WkgFetcher {
+    wkg_config: wasm_pkg_core::config::Config,
+    wkg_client_config: wasm_pkg_client::Config,
+    cache: FileCache,
+}
+
+impl WkgFetcher {
+    pub fn new(
+        wkg_config: wasm_pkg_core::config::Config,
+        wkg_client_config: wasm_pkg_client::Config,
+        cache: FileCache,
+    ) -> Self {
+        Self {
+            wkg_config,
+            wkg_client_config,
+            cache,
+        }
+    }
+
+    /// Load a `WkgFetcher` from a `CommonPackageArgs` and a `wasm_pkg_core::config::Config`
+    pub async fn from_common(
+        common: &CommonPackageArgs,
+        wkg_config: wasm_pkg_core::config::Config,
+    ) -> Result<Self> {
+        let cache = common
+            .load_cache()
+            .await
+            .context("failed to load wkg cache")?;
+        let wkg_client_config = common
+            .load_config()
+            .await
+            .context("failed to load wkg config")?;
+        Ok(Self::new(wkg_config, wkg_client_config, cache))
+    }
+
+    pub fn get_config(&self) -> &wasm_pkg_core::config::Config {
+        &self.wkg_config
+    }
+
+    /// Use configuration to create a caching client for fetching wkg dependencies
+    pub fn into_client(self) -> CachingClient<FileCache> {
+        let client = wasm_pkg_client::Client::new(self.wkg_client_config);
+        CachingClient::new(Some(client), self.cache)
+    }
+
+    /// Enable extended pull configurations for wkg config. Call before calling `Self::get_client` to
+    /// update configuration used.
+    pub async fn resolve_extended_pull_configs(
```

Comment:
Nit: Would it be better to use this as a builder or something similar rather than a function you have to remember to call?

Honestly this probably isn't worth doing now, but we might want to refactor this so it is clearer that you're essentially layering in extra config before calling `into_client`

Reply:
Resolving so this can merge